xayn-ai/src/ranker/context.rs

```diff
@@ -1,119 +1,231 @@
+/// Helper struct for [`find_closest_cois`].
+struct ClosestCois {
+    /// The ID of the positive centre of interest
+    pos_id: CoiId,
+    /// Distance from the positive centre of interest
+    pos_distance: f32,
+    pos_last_view: SystemTime,
+
+    /// Distance from the negative centre of interest
+    neg_distance: f32,
+    neg_last_view: SystemTime,
+}
 
-        Self { pos_avg, neg_max }
+fn find_closest_cois(
+    embedding: &Embedding,
+    user_interests: &UserInterests,
+    neighbors: usize,
+) -> Option<ClosestCois> {
+    let (pos_coi, pos_distance) = find_closest_coi(&user_interests.positive, embedding, neighbors)?;
+    let (neg_coi, neg_distance) = find_closest_coi(&user_interests.negative, embedding, neighbors)?;
+
+    ClosestCois {
+        pos_id: pos_coi.id(),
+        pos_distance,
+        pos_last_view: pos_coi.stats.last_view,
+        neg_distance,
+        neg_last_view: neg_coi.last_view,
     }
+    .into()
+}
+
+fn compute_score_for_embedding(
+    embedding: &Embedding,
+    user_interests: &UserInterests,
+    relevances: &mut RelevanceMap,
+    neighbors: usize,
+    horizon: Duration,
+    now: SystemTime,
+) -> Result<f32, Error> {
+    let cois = find_closest_cois(embedding, user_interests, neighbors)
+        .ok_or(Error::FailedToFindTheClosestCois)?;
+
+    let pos_decay = compute_coi_decay_factor(horizon, now, cois.pos_last_view);
+    let neg_decay = compute_coi_decay_factor(horizon, now, cois.neg_last_view);
+
+    relevances.compute_relevances(&user_interests.positive, horizon);
+    let pos_coi_relevance = relevances
+        .relevance_for_coi(&cois.pos_id).unwrap(/* we calculate relevance for all positive cois one line above */);
 
-    /// Calculates score from given positive distance and negative distance.
-    /// Both positive and negative distance must be >= 0.
-    pub(crate) fn calculate_score(&self, pos: f32, neg: f32) -> f32 {
-        debug_assert!(pos >= 0. && neg >= 0.);
-        let frac_pos = (self.pos_avg > 0.)
-            .then(|| (1. + pos / self.pos_avg).recip())
-            .unwrap_or(1.);
-        let frac_neg = (1. + (self.neg_max - neg)).recip();
+    let score = cois.pos_distance * pos_decay + pos_coi_relevance - cois.neg_distance * neg_decay;
 
-        (frac_pos + frac_neg) / 2.
+    Ok(score)
+}
+
+fn has_enough_cois(
+    user_interests: &UserInterests,
+    min_positive_cois: u32,
+    min_negative_cois: u32,
+) -> Result<(), Error> {
+    let pos_cois = user_interests.positive.len() as u32;
+    if pos_cois < min_positive_cois {
+        return Err(Error::NotEnoughPositiveCois {
+            expected: min_positive_cois,
+            has: pos_cois,
+        });
     }
+
+    let neg_cois = user_interests.negative.len() as u32;
+    if neg_cois < min_negative_cois {
+        return Err(Error::NotEnoughNegativeCois {
+            expected: min_negative_cois,
+            has: neg_cois,
+        });
+    }
+    Ok(())
+}
```

Comment:
do we need the detailed error information? otherwise we could simplify this to
```suggestion
fn has_enough_cois(
    user_interests: &UserInterests,
    min_positive_cois: u32,
    min_negative_cois: u32,
) -> bool {
    user_interests.positive.len() >= min_positive_cois && user_interests.negative.len() >= min_negative_cois
}
```
and return a simplified error at the call site like `Error::NotEnoughCois`.

Reply:
I've added the infomation because I thought it might help with debugging, but yeah it might be too much. We can add this back if we later find it helpful.

Reply:
done